### PR TITLE
[MPMD-318] Fix the aux classpath when includeTests is set to true

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/MPMD-318

In this PR I address the issue with auxiliary classpath that is being created in PmdReport.java. In case if 'includeTests' parameter is set to `true`, only test dependencies jars and test compiled classes are being added to the classpath, resulting in NoClassDefFoundErrors 

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MPMD) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue.  Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MPMD-XXX] - Subject of the JIRA Ticket`,
       where you replace `MPMD-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/pmd/PmdReport.java [hunk 1]
```diff
@@ -546,7 +548,7 @@ private String determineAuxClasspath() throws MavenReportException
     {
         try
         {
-            List<String> classpath = new ArrayList<>();
+            Set<String> classpath = new HashSet<>();
```
Comment (adangel) [inline src/main/java/org/apache/maven/plugins/pmd/PmdReport.java:551]: This should stay a list - the order is important for the classpath

File: src/main/java/org/apache/maven/plugins/pmd/PmdReport.java [hunk 2]
```diff
@@ -561,7 +563,7 @@ private String determineAuxClasspath() throws MavenReportException
                 }
                 TransformableFilter filter = new AndFilter( Arrays.asList(
                         new ExclusionsFilter( exclusionPatterns ),
-                        includeTests ? ScopeFilter.including( "test" ) : ScopeFilter.including( "compile" )
+                        includeTests ? ScopeFilter.including( "compile", "test" ) : ScopeFilter.including( "compile" )
```
Comment (adangel) [inline src/main/java/org/apache/maven/plugins/pmd/PmdReport.java:566]: This looks like this is the actual fix. We probably should also consider here the scope "provided". The goal is, that we get the classpath that is used for compilation...

File: src/main/java/org/apache/maven/plugins/pmd/PmdReport.java [hunk 3]
```diff
@@ -577,13 +579,19 @@ private String determineAuxClasspath() throws MavenReportException
                         dependencies.add( resolvedArtifact.getArtifact().getFile().toString() );
                     }
 
-                    List<String> projectCompileClasspath = includeTests ? localProject.getTestClasspathElements()
-                            : localProject.getCompileClasspathElements();
+                    List<String> projectClasspath = new ArrayList<>(
```
Comment (adangel) [inline src/main/java/org/apache/maven/plugins/pmd/PmdReport.java:582]: I think, this change is not needed: `getTestClasspathElements()` contains already `getCompileClasspathElements()` - so we would have e.g. `target/classes` now twice on the classpath

**Review by adangel — COMMENTED** [review]
Thanks for the PR.

As the relevant fix is within the code section, that is only used in aggregate mode, I guess, you are using `<aggregate>true</aggregate>`. Is this correct? Please consider then this FAQ entry: https://maven.apache.org/plugins/maven-pmd-plugin/faq.html#typeresolution_aggregate (and this issue https://issues.apache.org/jira/browse/MPMD-283).

I'll add a integration test though - right now, I don't see that *any* dependency is added to the auxclasspath when "includeTests" is used. That means, not only compile dependencies are missing, but also test dependencies....

File: src/main/java/org/apache/maven/plugins/pmd/PmdReport.java (hunk 2 above)
Comment (adangel) [reply src/main/java/org/apache/maven/plugins/pmd/PmdReport.java:566]: The missing provided scope might explain https://issues.apache.org/jira/browse/MPMD-317

**Comment by jira-importer** [comment]
Resolve #289

**Pull request merged**